
First, when you read in your files, they need to be read from your local folders(current working directory), or from an URL. Your folder does not include the original fang or snp files, so others cannot run the remaining of your scripts.

Your files should be read in as tsv instead of csv, the original files are not comma seperated.

When arranging your files by increasing/decreasing position, you could try doing it before seperating by chromosomes, so that you don't have to retype the arrange command multiple times.

#When merging your files, you could try just use: 
```{r}
teosinte_merge_file <- merge(SNPs_position_file,teosinte_transpose_data). 
```

For the decreasing position files, your missing values should be encoded by "-/-". You can check on Slack for the specific script for it. 

I didn't get the reason why you need to rename the SNPS_ID by Row.names. Also that command throws an error when I try to run it. 

For your first graph: file_plot_count, the script throws an error as following:
#Error in count(genotypes_merge_file, Chromosome) : object 'Chromosome' not found


I was not able to generate your first plot because of this error. 

#For the distribution plot of snps on each chromosome
you could try using geom_density instead of geom_point
```{r}
geom_density(aes(fill = as.character(Chromosome))) 
```
after your 
```{r}
ggplot(File_plot_dist, aes(x=Chromosome, y=Position))
```
this might give you a better visualization. 


#For the homozygous/heterozygous SNPs plot, in order to replace the characters, a simpler way might be as following:
```{r}
Genotype_files_new [Genotype_files_new %in% c("A/A","T/T","C/C","G/G")] <- 1
Genotypes_files_new [Genotype_files_new %in% c("A/G","A/T","C/G","C/T","G/T","A/C")] <- 0
Genotypes_files_new [Genotype_files_new %in% c("?/?")] <-2
```

Your homozygous/heterozygous SNPs plot looks great!
You also need to visualize another feature of the data, the choice is up to you!


When you write out your output files to either maize or teosinte subfolders, you need to specify the location of your output. such as following:

```{r}
write_tsv("./maize_files/maize_chr1_increasing.txt")
```

I noticed that you did not push to your online repository. You could try following in order to link your git repository with your local R_project:
first, create a online repository, second, create a R-project in your R-studio and choose Version control as your option. After that, you can put your online repository link in. Then R-project then should be linked to your local R-project. You can make changes and commit and push as necessary then. 




